File: TrippieBackend/TrippieBackend/Models/DTOs/Flights/FlightDto.cs
using TrippieBackend.Models.Enums;
using TrippieBackend.Models.DTOs;

namespace TrippieBackend.Models.DTOs.Flights;

public class FlightDto
{
    public Guid Id { get; set; }
    
    public TravelDirectionEnum TravelDirection { get; set; }
    
    public string? FlightNumber { get; set; }
    
    public string DepartureIataCode { get; set; } = string.Empty;
    
    public string DepartureCityName { get; set; } = string.Empty;
    
    public string ArrivalIataCode { get; set; } = string.Empty;
    
    public string ArrivalCityName { get; set; } = string.Empty;

    public AirportDto Arrival { get; set; } = null!;

    public AirportDto Departure { get; set; } = null!;

    public DateTime? DepartureTime { get; set; }
    
    public DateTime? ArrivalTime { get; set; }
}

File: trippie_frontend/lib/features/trip/data/flight_dto.dart
class AirportDto {
  const AirportDto({
    required this.id,
    required this.name,
    required this.city,
    required this.country,
    required this.iataCode,
  });

  final String id;
  final String name;
  final String city;
  final String country;
  final String iataCode;

  factory AirportDto.fromJson(Map<String, dynamic> json) => AirportDto(
        id: json['id'] as String,
        name: json['name'] as String,
        city: json['city'] as String,
        country: json['country'] as String,
        iataCode: json['iataCode'] as String,
      );

  String get displayName => '$iataCode — $city';
}

class FlightDto {
  const FlightDto({
    required this.id,
    required this.travelDirection,
    this.flightNumber,
    required this.departure,
    required this.arrival,
    this.departureTime,
    this.arrivalTime,
  });

  final String id;
  final String travelDirection; // OUTBOUND / RETURN
  final String? flightNumber;
  final AirportDto departure;
  final AirportDto arrival;
  final DateTime? departureTime;
  final DateTime? arrivalTime;

  factory FlightDto.fromJson(Map<String, dynamic> json) => FlightDto(
        id: json['id'] as String,
        travelDirection: json['travelDirection'] as String,
        flightNumber: json['flightNumber'] as String?,
        departure: AirportDto.fromJson(json['departure'] as Map<String, dynamic>),
        arrival: AirportDto.fromJson(json['arrival'] as Map<String, dynamic>),
        departureTime: json['departureTime'] != null
            ? DateTime.parse(json['departureTime'] as String)
            : null,
        arrivalTime: json['arrivalTime'] != null
            ? DateTime.parse(json['arrivalTime'] as String)
            : null,
      );

  String get routeLabel =>
      '${departure.iataCode} → ${arrival.iataCode}';

  bool get isOutbound => travelDirection == 'OUTBOUND';

  String get departureIataCode => departure.iataCode;
  String get arrivalIataCode => arrival.iataCode;
  String get departureCityName => departure.city;
  String get arrivalCityName => arrival.city;
}